Improving painting methods, especially for drawing

asmanur opened the pull request:
This patch modifies the repainting discipline to allow smoothless drawing. There is now two painting discipline:

- The standard one, which behaves like the previous one but also forces a repaint every 500ms.

- The fast one, only used when drawing which moreover repaints updates in the direct mode, which is very fast but can only paint it black.

- Moreover, when the pen is released, we also update the screen so that the user can see the whole stroke.

This allows for drawing on shared whiteboard with a black pen with reasonable latency.

Disclaimer: I have not written a line of C++ in about a decade so I'm not really up-to-date with coding standards (not that I ever was).

matteodelabre commented on src/app/screen.cpp:
```diff
@@ -40,31 +47,65 @@ screen::screen(rmioc::screen& device, rfbClient* vnc_client)
+                this->repaint(true);
+
             // Wait until the next update is due
             return {/* quit = */ false, /* timeout = */ remaining_wait_time};
```
Shouldn’t this be changed to

```suggestion
            return {/* quit = */ false, /* timeout = */ std::min(remaining_wait_time, must_repaint)};
```

So that the next update is scheduled as soon as one of the two conditions are met?

asmanur replied:
True, I did not spot out this !

matteodelabre replied:
Thanks!

matteodelabre commented on src/app/pen.cpp:
```diff
@@ -38,12 +39,19 @@ auto pen::process_events() -> event_loop_status
+                : MouseButton::None;
+            this->send_button_press(screen_x, screen_y, new_state);
+
+            if (this->state != new_state) {
+                if (this->state == MouseButton::Left)
+                    this->screen.set_repainting_mode(repainting_mode::fast);
+                else
+                    this->screen.set_repainting_mode(repainting_mode::standard);
+            }
+            this->state = new_state;
```
Can we simplify this to:

```suggestion
            if (new_state == MouseButton::Left)
            {
                this->screen.set_repainting_mode(repainting_mode::fast);
            }
            else
            {
                this->screen.set_repainting_mode(repainting_mode::standard);
            }
```

Which eliminates the need for the `state` member variable? Or did I miss something?

asmanur replied:
Yes, you are right I forgot something: when you release the pen you want the screen to be redrawn at that moment to see your whole stroke (at least in my opinion). I have added this in the patch below.

matteodelabre replied:
Makes sense, thanks!

matteodelabre reviewed (commented):
Thanks @asmanur, this looks good! I’m curious how the (1) `last_update_time + update_delay ⩾ now` and (2) `last_repaint_time + update_max_delay ⩾ now` conditions interact and whether one is redundant with the other.

My first intuition would be that (1) is never reached when `update_delay > update_max_delay` (which is not the case in the code). Otherwise, if `update_delay ⩽ update_max_delay`, (1) is reached after each batch of updates and triggers a repaint before (2) is met, which saves `update_max_delay − update_delay` milliseconds.

asmanur commented:
Thanks for the comments! I modified 
17c6bb4 with std::min and added 88bd6cd for repainting on pen release.

As for the two methods, yes I did it that way to implement what was discussed in the Issue. But I am not sure the first one is still necessary - maybe it's better for usage such as text editing where the buffer does end up stabilising?

matteodelabre commented on src/app/pen.cpp:
```diff
@@ -38,12 +39,21 @@ auto pen::process_events() -> event_loop_status
+            this->send_button_press(screen_x, screen_y, new_state);
+
+            if (this->state != new_state) {
+                if (this->state == MouseButton::Left)
```
I believe this should be changed to

```suggestion
                if (new_state == MouseButton::Left)
```

Otherwise the condition is inverted since `this->state` contains the previous state at this point.

asmanur replied:
Well spotted out!

Fixed the commit.

matteodelabre commented:
Perfect, thanks a lot! I’ll make a new release including this change soon.
